core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java

```diff
@@ -0,0 +1,165 @@
+ * public class someClass {
+ *   public void doSomething() {
+ *     // Do the work of the method. One result is a measurement of something.
+ *     int measurement = doSomeWork();
+ *     // Make an update to the class state, based on this measurement. This work can take some
+ *     // time, but can be done asynchronously, in the background.
+ *     update(measurement);
+ *   }
+ *
+ *   public void update(int arg) {
+ *     // do something
+ *   }
+ * }
+ * </pre>
+ *
+ * <p>The work of calling {@code someClass.update()} can be executed in the backgound as follows:
+ * <pre>
+ * public class someClass {
+ *   // Add a EventQueueEntry class that will process the update call.
+ *   private static final class SomeClassUpdateEvent implements EventQueueEntry {
+ *     private final SomeClass someClassInstance;
+ *     private final int arg;
+ *
+ *     SomeObjectUpdateEvent(SomeObject someClassInstance, int arg) {
+ *       this.someClassInstance = someClassInstance;
+ *       this.arg = arg;
+ *     }
+ *
+ *     &#064;Override
+ *     public void process() {
+ *       someClassInstance.update(arg);
+ *     }
+ *   }
+ *
+ *   public void doSomething() {
+ *     int measurement = doSomeWork();
+ *     // Instead of calling update() directly, create an event to do the processing, and insert
+ *     // it into the EventQueue. It will be processed in a background thread, and doSomething()
+ *     // can return immediately.
+ *     EventQueue.getInstance.enqueue(new SomeClassUpdateEvent(this, measurement));
+ *   }
+ * }
+ *</pre>
+ */
+
+@ThreadSafe
+public final class EventQueue {
+  // An event in the {@link EventQueue}. Just holds a reference to an EventQueueEntry.
+  private static final class InstrumentationEvent {
+    private EventQueueEntry entry = null;
+
+    // Set the EventQueueEntry associated with this InstrumentationEvent.
+    void setEntry(EventQueueEntry entry) {
+      this.entry = entry;
+    }
+
+    // Get the EventQueueEntry associated with this InstrumentationEvent.
+    EventQueueEntry getEntry() {
+      return entry;
+    }
+  }
+
+  // Factory for InstrumentationEvent.
+  private static final class InstrumentationEventFactory
+      implements EventFactory<InstrumentationEvent> {
+    @Override
+    public InstrumentationEvent newInstance() {
+      return new InstrumentationEvent();
+    }
+  }
+
+  /**
+   * Every event that gets added to {@link EventQueue} will get processed here. Just calls the
+   * underlying process() method.
+   */
+  private static final class InstrumentationEventHandler
+      implements EventHandler<InstrumentationEvent> {
+    @Override
+    public void onEvent(InstrumentationEvent event, long sequence, boolean endOfBatch) {
+      event.getEntry().process();
+    }
+  }
+
+  // The single instance of the class.
+  private static final EventQueue eventQueue = new EventQueue();
+  // The event queue is built on this {@link Disruptor}.
+  private final Disruptor<InstrumentationEvent> disruptor;
```

Comment:
interestingly this type isn't used, yet (and the core-impl package is otherwise little code), so maybe just keep in mind that as the impl part grows, there will be parts others can harvest. guava is usually used by others pretty well understood etc. disruptor has a relatively low change frequency at the moment, and likely stable enough to have a core dep, but it is less understood.

Probably only async folks (who also use disruptor) who want to use this impl will need to make a custom build which shades it and publishes a variant which doesn't conflict. We can always wait until they complain (which may never happen)